### DISCUSSION TITLE
Uploading attachments to salesforce

### DISCUSSION THREAD

**Discussion by opike** [jsforce/jsforce#1426] [Ideas]
This thread describes how to download Attachments:
https://github.com/jsforce/jsforce/issues/19

Does anyone have an example that shows how to do an upload?

**Comment by brianchenault** [comment]
I'm actually trying to get this working as well. It doesn't seem as if the jsforce Chatter module allows for multipart requests.

**Comment by stomita** [comment]
Attaching a new attachment is not implemented right now. If you already have an attachment, you can access the attachment by following code

```
var attachStream = conn.sobject('Attachment').record(attachmentId).blob('Body');
var outFileStream = fs.createWriteStream('/path/to/file')
attachStream.pipe(outFileStream)
```

Planing to implement the attachment features as well as chatter in future release.

**Comment by idanblich** [comment]
Hi , 
Any update regarding Attaching a new attachment ?
TX.
Idan

**Comment by yawetse** [comment]
uploading attachments:

``` javascript
'use strict';
var fileOnServer = '/home/ubuntu/myuploaddir/randomfile/randomimage.jpg',
    fileName = 'MyRandomImage.jpg',
    fileType = 'image/jpeg';

fs.readFile(fileOnServer, function (err, filedata) {
    if (err){
        console.error(err);
    }
    else{
        var base64data = new Buffer(filedata).toString('base64');
        jsForceConn.sobject('Attachment').create({ 
                ParentId: 'mysalesforceContactID',
                Name : fileName,
                Body: base64data,
                ContentType : fileType,  
            }, 
            function(err, uploadedAttachment) {
                console.log(err,uploadedAttachment);
        });
}
});
```

**Comment by stomita** [comment]
Neat, @yawetse !
We may add a stream version of attachment upload functionality in future version.

**Comment by yawetse** [comment]
that would be pretty sweet

**Comment by oaaraya** [comment]
Attaching to Chatter would be a great feature.

**Comment by mvigano** [comment]
Hi @stomita,
I'm trying to upload a 6.3MB pdf file as an attachment to a custom object as suggested by @yawetse...but I get my node.js process being killed on the create(...).
The create call should restrict attachment files to a maximum size of 25 MB, shouldn't it?
Any advice?
Any update on the multipart/form-data streaming version?
Thanks,
Marcello

**Comment by yawetse** [comment]
@mvigano can you include a code snippet and your console output?

Also are you trying to stream/pipe the file directly from the upload or are you saving the upload to disk first?

I've only gotten it to work by saving the uploaded file to disk first and then uploading to Salesforce

**Comment by mvigano** [comment]
Thank you for your prompt reply, @yawetse !
I'm currently fetching my attachments using request: 

```
var request = require('request').defaults({ encoding: null })
```

and right uploading them to SFDC without saving them first to disk...

It seems that everything works fine until the file size doesn't exceed 5 MB....

Here is it:

``` javascript
step(
    function fetchAttachments() {
        var group = this.group();
        attachments.forEach(function(attach, i) {
            logger.debug('Fetching attachment', attach.url);
            request(attach.url, group());   
        });         
    },
    function uploadAttachToSFDC(err, files) {
        logger.debug('files:', files);
        if (err) { throw err; }
        if (files && _.isArray(files) ) {               
            var group = this.group();
            files.forEach(function(file, i) {
                if (file.statusCode == 200) {
                    logger.debug('Attachment ' + attachments[i].url + ' fetched');                  
                        // Maximum file size limit to 5 MB
                        if (parseInt(file.headers["content-length"], 10) < 5000000 ) {
                            var base64data = new Buffer(file.body).toString('base64'); 
                            // Upload attachment to SFDC
                            conn.sobject('Attachment').create({
                            ParentId : parentId,
                                Name: attachments[i].meta.originalName,
                                Body: base64data,
                                ContentType: file.headers["content-type"], // attachments[i].meta.mimetype,
                                //OwnerId: '',
                                Description: attachments[i].meta.desc 
                            }, group());
                        } else {
                            logger.error('File size exceeding 5MB. Attachment NOT uploaded to SFDC');
                        }                                   
                }
            });
        }
    },
...

```

And this is my console output:
<img width="922" alt="screen shot 2015-10-08 at 16 46 50" src="https://cloud.githubusercontent.com/assets/6479244/10369772/3e660e30-6ddc-11e5-93b3-d972b94b6993.png">

**Comment by yawetse** [comment]
@mvigano interesting, I haven't come across that limitation with the way I implemented it in the past. 

Are there any restrictions on the salesforce side (does your salesforce org have file limit constraints?)

Also side note, I've never used step before but async and bluebird (for control flow) worked well for me. It's probably unrelated to the file upload limitation, unless there's some sort of memory constraint i'm unaware of.

**Comment by mvigano** [comment]
@yawetse ,
Thank you very much for looking into this. I'm using my Salesforce Developer Account...
Just tried saving the file to disk first, but I get my proc killed anyway.....

**Comment by krishan1kamal** [comment]
What is parentId in upload attachment?

**Comment by paustint** [comment]
Please refer to the Salesforce documentation for information about their
API parameters. The ParentId is the record the attachment is being added
to. Your SObject must have notes and attachments enabled.

Check here for more details:
https://help.salesforce.com/articleView?id=loader_attachments.htm&type=0

And search for the API documentation for more technical information.

**Comment by vascoV** [comment]
Hey, guys,

I am using Salesforce Mobile SDK (Android) to upload an image. I am using the SDriveTools app which integrates Amazon S3 as a related list.

So I uploaded the image to the S3 amazon space using the Amazon SDK (Android), but I have an issue with the SDriveTools app which integrated S3 as related List into the salesforce as mentioned above.

Any ideas?
